### ISSUE TITLE
Empty class elements left after swap

### ISSUE THREAD

**Issue by dmcg** [bigskysoftware/htmx#1701]
Inspecting the page after a swap, htmx leaves an empty `class=“”` attribute on the target element. It’s benign, but means that the resulting page is not the same as the a page for the same server state rendered from scratch, which makes testing harder.

 You can see the problem in https://youtu.be/qYuSr__OiHk at 23:15

**Comment by Telroshan** [comment]
Hey, that's because htmx adds temporary [classes](https://htmx.org/reference/#classes) such as `htmx-request` to your elements while the requests are in flight, to let you use some styling when needed, such as a [`hx-indicator`](https://htmx.org/attributes/hx-indicator/) to display say a loading spinner until the request is done.

Since those classes are added, it's then actually simply because of how the JS API works that the empty `class` attribute remains, it's not related to htmx itself.

You can observe that behaviour in [this JSFiddle](https://jsfiddle.net/c9vjuwLh/)
```html
<div id="test">Inspect me in F12</div>

<button type="button" onclick="test.classList.add('test')">Add class</button>

<button type="button" onclick="test.classList.remove('test')">Remove class (classList)</button>
<button type="button" onclick="test.className=''">Remove class (className)</button>
```

Using both `classList` and `className`, you'll see that after removing the class, the `class` attribute doesn't go away, and remains here, empty

**Comment by dmcg** [comment]
I can see how it happens, and if it can’t be fixed then so be it, but it seems a shame that testing is made more complicated than it could be?

**Comment by duanemoody** [comment]
The JSFiddle shows the code leaving behind `class=""` instead of a blank `class` attribute for both RemoveClass buttons, so something different's happening here.

For htmx, the code temporarily creating a `class` attribute for the purpose of handling a trigger could check whether that attribute already exists on the element, and if not set a temporary element like `data-cleanup="class"` for htmx to check for, then have`element.classList.length == 0 && element.removeAttribute('class');` in whatever callback these functions use. 

Assuming a nonboolean attribute with no value is equivalent to no attribute at all is only true to an HTML renderer, but not necessarily to other DOM-aware code running on the page.

**Comment by duanemoody** [comment]
> Hey, that's because htmx adds temporary [classes](https://htmx.org/reference/#classes) such as `htmx-request` to your elements while the requests are in flight, to let you use some styling when needed, such as a [`hx-indicator`](https://htmx.org/attributes/hx-indicator/) to display say a loading spinner until the request is done.
> 
> Since those classes are added, it's then actually simply because of how the JS API works that the empty `class` attribute remains, it's not related to htmx itself.
> 
> You can observe that behaviour in [this JSFiddle](https://jsfiddle.net/c9vjuwLh/)
> 
> ```
> <div id="test">Inspect me in F12</div>
> 
> <button type="button" onclick="test.classList.add('test')">Add class</button>
> 
> <button type="button" onclick="test.classList.remove('test')">Remove class (classList)</button>
> <button type="button" onclick="test.className=''">Remove class (className)</button>
> ```
> 
> Using both `classList` and `className`, you'll see that after removing the class, the `class` attribute doesn't go away, and remains here, empty

This is true, but htmx's `removeClassFromElement` method, which is supposed to check for and remove empty class attributes, has a logic branch that skips the check/remove statement if `delay` is defined (not sure why) and I'm guessing the decision to not rename `delay` inside functions is resulting in a scope problem where `if (delay)` is actually evaluating against a global instead of a passed argument.

**Comment by duanemoody** [comment]
> Hey, that's because htmx adds temporary [classes](https://htmx.org/reference/#classes) such as `htmx-request` to your elements while the requests are in flight, to let you use some styling when needed, such as a [`hx-indicator`](https://htmx.org/attributes/hx-indicator/) to display say a loading spinner until the request is done.
Looking at the source code, it's actually because `doSettle` is for some reason still manually adding/removing those utility classes to the elements through bare-metal `elt.classList.add`/`elt.classList.remove` statements, instead of leveraging its own `addClassToElement`/`removeClassFromElement` methods, which _do_ bother doing checks on `elt.classList.length === 0` then removing the attribute as needed.

Instead of second-guessing the reason for not employing those higher-level methods, I just added the same check to `doSettle` after `elt.classList.remove` in my last pull request which seems to have solved the problem.
